### PR TITLE
Update cluster/gce/ scripts to support Windows nodes.

### PR DESCRIPTION
**Pull request by pjh**
/kind feature

**What this PR does / why we need it**: This PR introduces support for Windows nodes into the cluster bringup scripts for GCE. [Windows node support](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190103-windows-node-support.md) is targeting Kubernetes 1.14 for GA.

There are two main parts to the PR: changes to the existing scripts under `cluster/gce` and the addition of several PowerShell scripts that run on the Windows nodes to connect them to the cluster. These scripts currently work against Windows Server version 1803.

A version of these changes is already being tested at https://testgrid.k8s.io/google-windows#windows-prototype. Another test job (pending https://github.com/kubernetes/test-infra/pull/11010) targeting these changes will be added once the PR is merged.

The full history of these changes can be seen at https://github.com/pjh/kubernetes/tree/windows-up. Thanks @yujuhong and @mtaufen for their contributions.

```release-note
Introduced support for Windows nodes into the cluster bringup scripts for GCE.
```

### CODE REVIEW

**Comment by yujuhong** [comment]
The PR is huge, but the majority of it is adding a new, isolated `cluster/gce/win1803` package.

/cc @roberthbailey @zmerlynn @mikedanese 
Does any of you have time to time to help review the changes in `cluster/` and `cluster/gce/`?

**Comment by pjh** [comment]
CLA signed and commit amended with proper e-mail address.

**Comment by pjh** [comment]
CC some folks from sig-windows: @PatrickLang @michmike @benmoss @astrieanna

**Review by michmike — COMMENTED** [review]
why is 1803 the OS of choice? SIG-Windows plans to GA with windows node in kubernetes v1.14, but we will only support windows server 2019 (win1809)

File: cluster/gce/config-default.sh [hunk 1]
```diff
@@ -467,3 +476,7 @@ ENABLE_NODE_TERMINATION_HANDLER="${ENABLE_NODE_TERMINATION_HANDLER:-false}"
 if [[ "${NODE_TERMINATION_HANDLER_IMAGE:-}" ]]; then
   PROVIDER_VARS="${PROVIDER_VARS:-} NODE_TERMINATION_HANDLER_IMAGE"
 fi
+
+# Taint Windows nodes by default to prevent Linux workloads from being
+# scheduled onto them.
+WINDOWS_NODE_TAINTS="${WINDOWS_NODE_TAINTS:-node.kubernetes.io/os=windows:NoSchedule}"
```
Comment (michmike) [inline cluster/gce/config-default.sh]: We would prefer if the taint was specific to the operating system of windows since there has to be a direct correllation between host OS version and container OS version.
so the taint should be `Windows1803` in this case.
Note that 1803 is not supported by sig-windows for GA

File: cluster/gce/config-test.sh [hunk 2]
```diff
@@ -486,3 +496,7 @@ ENABLE_NODE_TERMINATION_HANDLER="${ENABLE_NODE_TERMINATION_HANDLER:-false}"
 if [[ "${NODE_TERMINATION_HANDLER_IMAGE:-}" ]]; then
   PROVIDER_VARS="${PROVIDER_VARS:-} NODE_TERMINATION_HANDLER_IMAGE"
 fi
+
+# Taint Windows nodes by default to prevent Linux workloads from being
+# scheduled onto them.
+WINDOWS_NODE_TAINTS="${WINDOWS_NODE_TAINTS:-node.kubernetes.io/os=windows:NoSchedule}"
```
Comment (michmike) [inline cluster/gce/config-test.sh]: same comment as above about the taint name

**Comment by michmike** [comment]
aside from a couple of comments i made above, i am ok with this PR. once i hear back on my comments, i can LGTM

**Comment by benmoss** [comment]
Doesn't this [deprecation notice](https://github.com/kubernetes/kubernetes/tree/2d6834b539dd434f70060ba3bc2d11d0465f12da/cluster#deprecation-notice-this-directory-has-entered-maintenance-mode-and-will-not-be-accepting-new-providers-deployments-in-this-directory-will-continue-to-be-maintained-and-supported-at-their-current-level-of-support) suggest against making this kind of improvement to these scripts? I don't mean to discourage this work, I have been wishing for more of the standard Kubernetes tooling to start supporting bringing up Windows nodes, but I thought the intention was to focus more effort on out-of-tree tools like kubespray, kops, etc.

**Comment by yujuhong** [comment]
> why is 1803 the OS of choice? SIG-Windows plans to GA with windows node in kubernetes v1.14, but we will only support windows server 2019 (win1

This is not the choice for 1.14, but we've been testing 1803 for a while and wanted to get this upstream AND THEN switch to 2019 to ensure stability.

As discussed in the last sig-windows meeting, there are issues with 2019 that are being patched, so we'll qualify and switch to 2019 later.

**Comment by yujuhong** [comment]
> Doesn't this [deprecation notice](https://github.com/kubernetes/kubernetes/tree/2d6834b539dd434f70060ba3bc2d11d0465f12da/cluster#deprecation-notice-this-directory-has-entered-maintenance-mode-and-will-not-be-accepting-new-providers-deployments-in-this-directory-will-continue-to-be-maintained-and-supported-at-their-current-level-of-support) suggest against making this kind of improvement to these scripts? I don't mean to discourage this work, I have been wishing for more of the standard Kubernetes tooling to start supporting bringing up Windows nodes, but I thought the intention was to focus more effort on out-of-tree tools like kubespray, kops, etc.

Yes, but we (GCP) still rely on these scripts to for the majority of our test jobs in the testgrid. It's unreasonable to block this project until the replacement is completely ready.
Please also be aware that the majority of the changes reside in the `gce` package.

**Comment by yujuhong** [comment]
> @spiffxp was also looking at becoming an owner for the cluster directory. Maybe he wants to take a first pass?

/assign @spiffxp 

@spiffxp please unassign if you don't have bandwidth to review this. Thanks!

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (yujuhong) [reply cluster/gce/config-default.sh]: Yes, I'm aware of this. That's why I brought this up in the Windows GA KEP :-)
I'd like some clear recommendation of how OS version matching is going to work. For example, should it be win1809 or win2019? Is `1809` going to be compatible with `1903`? I think this may eventually lead to more confusion to the users.

BTW, this is a temporary taint we choose for GCE. We can (and will) change this later. The main purpose of this PR is to get the code in tree, so we can start testing against k8s and make incremental changes.

File: cluster/gce/config-test.sh (hunk 2 above)
Comment (yujuhong) [reply cluster/gce/config-test.sh]: See reply in https://github.com/kubernetes/kubernetes/pull/73442#discussion_r251932402

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -342,7 +364,8 @@ function upload-tars() {
   fi
 }
 
-# Detect minions created in the minion group
+# Detect minions created in the minion group. Finds both Linux and Windows
+# minion nodes.
```
Comment (spiffxp) [inline cluster/gce/util.sh]: Nit that has nothing to do with the change in this PR, but since you're changing this line: can we stop referring to minions. eg:

```Detect Linux and Windows nodes created in the instance group.```

File: cluster/gce/util.sh [hunk 4]
```diff
@@ -695,60 +750,170 @@ function construct-kubelet-flags {
...
+    print-common-kubelet-config
+    if [[ "${master}" == "true" ]]; then
+      print-master-kubelet-config
+    else
+      print-common-node-kubelet-config
+      if [[ "${os}" == "linux" ]]; then
+        print-linux-node-kubelet-config
+      elif [[ "${os}" == "windows" ]]; then
+        print-windows-node-kubelet-config
+      fi
```
Comment (spiffxp) [inline cluster/gce/util.sh]: we should probably error out if `os` is neither `linux` nor `windows`

File: cluster/gce/util.sh [hunk 5]
```diff
@@ -695,60 +750,170 @@ function construct-kubelet-flags {
...
   - ${quoted_dns_server_ip}
 clusterDomain: ${quoted_dns_domain}
-staticPodPath: /etc/kubernetes/manifests
 readOnlyPort: 10255
 EOF
 
-    # --- begin master-specific config ---
-    if [[ "${master}" == "true" ]]; then
-        cat <<EOF
+  # Note: ENABLE_MANIFEST_URL is used by GKE
```
Comment (spiffxp) [inline cluster/gce/util.sh]: FWIW I'm not actually seeing this referenced anywhere in kubernetes/kubernetes nor kubernetes/test-infra today.  Can you either TODO to remove this, or add an additional commit that does

File: cluster/gce/util.sh [hunk 6]
```diff
@@ -1670,11 +1875,15 @@ function validate-node-local-ssds-ext(){
...
-# $3: String of comma-separated metadata entries (must all be from a file).
+# $3: String of comma-separated metadata-from-file entries.
+# $4: String of comma-separated metadata (key=value) entries.
+# $5: "linux" or "windows".
 function create-node-template() {
   detect-project
   detect-subnetworks
   local template_name="$1"
+  local metadata_values="$4"
+  local linux_or_windows="$5"
```
Comment (spiffxp) [inline cluster/gce/util.sh]: we called this `os` elsewhere (I think I like this name better)

again though we should error out if something other than "linux" or "windows" makes its way here

File: cluster/gce/util.sh [hunk 7]
```diff
@@ -1737,6 +1946,15 @@ function create-node-template() {
     "${ENABLE_IP_ALIASES:-}" \
     "${IP_ALIAS_SIZE:-}")
 
+  local node_image_flags=""
+  if [[ "${linux_or_windows}" == 'linux' ]]; then
+      node_image_flags="--image-project ${LINUX_NODE_IMAGE_PROJECT} --image ${LINUX_NODE_IMAGE}"
+  else
```
Comment (spiffxp) [inline cluster/gce/util.sh]: it's a string not a boolean let's make sure it's "windows"

**Review by spiffxp — COMMENTED** [review]
I don't have the bandwidth to suggest something better than the "linux and windows copies of everything" approach taken here.  So mostly some minor maintenance nits

Hopefully we're doing things a different way before we face having to do N=3 variants of nodes.

Not powershell literate so I can't comment there.

I couldn't really follow the history mentioned in the comment.  Splitting up into more than one commit would have helped.  I'm mostly trusting that this isn't breaking things by the PR jobs not breaking.

**Comment by spiffxp** [comment]
I don't know that I'll have the bandwidth to followup beyond this so feel free to /lgtm without me

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (pjh) [reply cluster/gce/config-default.sh]: I agree - for the taint name we can wait for the approach to be finalized in the KEP / elsewhere, then we can easily change this.

**Review by pjh — COMMENTED** [review]
@michmike I have a branch / commit where I've changed the default Windows version to 2019: https://github.com/pjh/kubernetes/commit/3a2db8921d5f50aab382c7280ff5ce15b8e1dce2 / https://github.com/pjh/kubernetes/tree/gce-windows-cluster-2019. I expect we'll merge that change soon, after this PR goes in and our test job against it stabilizes.

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (michmike) [reply cluster/gce/config-default.sh]: Microsoft publishes a clear compatibility guide here. https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility. we document this in the KEP and don't have any ownership or influence here. it is what it is. 1809 and 1903 are incompatible for Windows server containers.

the name of the taint is not as important as long as it is consistent between containers and container host and it describes the OS version it is compatible with.

**Comment by pjh** [comment]
> I don't have the bandwidth to suggest something better than the "linux and windows copies of everything" approach taken here. So mostly some minor maintenance nits

@spiffxp Thank you for the prompt review!

> Hopefully we're doing things a different way before we face having to do N=3 variants of nodes.
> 
> Not powershell literate so I can't comment there.
> 
> I couldn't really follow the history mentioned in the comment. Splitting up into more than one commit would have helped. 

Thanks for the feedback. Do you have any suggestions for how to split this PR into separate commits? If so I'd be glad to do the split to help reviewers.

I considered how to do this but I didn't really see any logical changes to split this into... as you noted, most of the changes to existing files just make "linux and windows copies of everything", so it makes sense to just do this all in one go.

I could put the powershell scripts into a separate commit but since the set of files in the commits would be disjoint I don't think that would achieve much.

> I'm mostly trusting that this isn't breaking things by the PR jobs not breaking.

Yeah. To avoid breaking things I tried to minimize the changes to the existing Linux variables/code, and avoided changing any "externally-visible" environment variable names to maintain compatibility for users who are setting those. If anyone has suggestions for how to further test the changes please let me know.

File: cluster/gce/util.sh [hunk 8]
```diff
@@ -57,23 +59,43 @@ fi
 
 # Sets node image based on the specified os distro. Currently this function only
 # supports gci and debian.
-function set-node-image() {
-    if [[ "${NODE_OS_DISTRIBUTION}" == "gci" ]]; then
+function set-linux-node-image() {
```
[the remark this answers is not in the document]
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 3 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 4 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 5 above)
Comment (pjh) [reply cluster/gce/util.sh]: TODO added for @mtaufen.

File: cluster/gce/util.sh (hunk 6 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 7 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

**Comment by yujuhong** [comment]
Thanks for reviewing!

> Hopefully we're doing things a different way before we face having to do N=3 variants of nodes.

This is my hope too, and I think cluster API, etc should be ready by then (if we ever do that).

> Not powershell literate so I can't comment there.

We mainly want a reviewer for the cluster/gce changes. We can get a separate powershell reviewer after the main change is okay'd by an approver. Those scripts are isolated and won't affect existing features.

**Comment by yujuhong** [comment]
@pjh could you add an OWNERS file for the `win1803` package? You can put my name as the approver for now.

**Comment by pjh** [comment]
> @pjh could you add an OWNERS file for the `win1803` package? You can put my name as the approver for now.

Done.

**Comment by yujuhong** [comment]
@pjh FYI, `hack/verify-shellcheck.sh` failed.

File: cluster/gce/util.sh (hunk 5 above)
Comment (mtaufen) [reply cluster/gce/util.sh]: @pjh thanks.

**Comment by pjh** [comment]
> @pjh FYI, `hack/verify-shellcheck.sh` failed.

Thanks, I've addressed all these failures.

**Comment by pjh** [comment]
Sadly, it looks like changing some of the "internal" environment variable names is more dangerous than I anticipated. For example, `cluster/gce/config-test.sh` sets this variable:

```
NODE_OS_DISTRIBUTION=${KUBE_NODE_OS_DISTRIBUTION:-${KUBE_OS_DISTRIBUTION:-gci}}
```

I thought it would be ok to change the "internal" name (within the `cluster/` scripts) of this variable to `LINUX_NODE_OS_DISTRIBUTION`, leaving the "external" name `KUBE_NODE_OS_DISTRIBUTION` unchanged to remain compatible with existing users. Unfortunately it turns out there are other, less-obvious users of the internal name too - `hack/ginkgo-e2e.sh` uses `NODE_OS_DISTRIBUTION` directly, not `KUBE_NODE_OS_DISTRIBUTION`. This is causing pull-kubernetes-e2e-gce to run tests against this PR that should be skipped.

Similarly, a quick grep for `NUM_NODES` in the entire tree turns up a million uses, so introducing `LINUX_NUM_NODES` instead and using it throughout `cluster/` is still potentially dangerous. It might be possible to hunt down all the uses in the tree but even then who knows what else is consuming these variables.

I'm working on updating the PR so that no existing variable names are changed at all. It's unfortunate that we'll have e.g. `NUM_NODES` and `NUM_WINDOWS_NODES`, instead of `NUM_LINUX_NODES` and `NUM_WINDOWS_NODES`.

**Comment by yujuhong** [comment]
> I'm working on updating the PR so that no existing variable names are changed at all. It's unfortunate that we'll have e.g. `NUM_NODES` and `NUM_WINDOWS_NODES`, instead of `NUM_LINUX_NODES` and `NUM_WINDOWS_NODES`.

Alternatively, we could still keep `NUM_NODES` as the number of total nodes, and using `NUM_WINDOWS_NODES` and `NUM_LINUX_NODES` for configuring nodes in `cluster/ `.
E.g.,
- if `NUM_NODES` is set to 5, cluster creates 5 linux nodes
- if `NUM_NODES` is set to 5, and `NUM_WINDOWS_NODES` is set to 2, the script automatically set `NUM_LINUX_NODES` to 3. 

Internally, everything uses `NUM_WINDOWS_NODES` and `NUM_LINUX_NODES`. Externally, `NUM_NODES` can be configured normally when `NUM_WINDOWS_NODES` is not set (defaut:0). 
What do you think?

**Comment by pjh** [comment]
> I'm working on updating the PR so that no existing variable names are changed at all. It's unfortunate that we'll have e.g. `NUM_NODES` and `NUM_WINDOWS_NODES`, instead of `NUM_LINUX_NODES` and `NUM_WINDOWS_NODES`.

This is done and I've tested that cluster bringup still works.

@michmike did you want to take another look at this?

**Comment by yujuhong** [comment]
/lgtm
/approve

@spiffxp did a pass of review.

@roberthbailey @zmerlynn @mikedanese, does anyone of you have time to review and/or approve?

File: cluster/gce/util.sh [hunk 9]
```diff
@@ -695,60 +755,175 @@ function construct-kubelet-flags {
...
   fi
-  if [[ -n "${MAX_PODS_PER_NODE:-}" ]]; then
-    flags+=" --max-pods=${MAX_PODS_PER_NODE}"
+
+  KUBELET_ARGS="${flags}"
+}
+
+# Sets KUBELET_ARGS with the kubelet flags for Windows nodes.
+function construct-windows-kubelet-flags {
+  local flags=$(construct-common-kubelet-flags)
```
Comment (mikedanese) [inline cluster/gce/util.sh]: quote command substitutions here and below:

https://google.github.io/styleguide/shell.xml?showone=Quoting#Quoting

File: cluster/gce/util.sh [hunk 10]
```diff
@@ -695,60 +755,175 @@ function construct-kubelet-flags {
...
+
+  # Turn off kernel memory cgroup notification.
+  flags+=" --experimental-kernel-memcg-notification=false"
+
   KUBELET_ARGS="${flags}"
 }
 
 # $1: if 'true', we're rendering config for a master, else a node
 function build-kubelet-config {
   local master=$1
```
Comment (mikedanese) [inline cluster/gce/util.sh]: Quote argument variables.

File: cluster/gce/util.sh [hunk 11]
```diff
@@ -950,6 +1136,10 @@ $(echo "${CUSTOM_CALICO_NODE_DAEMONSET_YAML:-}" | sed -e "s/'/''/g")
 CUSTOM_TYPHA_DEPLOYMENT_YAML: |
 $(echo "${CUSTOM_TYPHA_DEPLOYMENT_YAML:-}" | sed -e "s/'/''/g")
 EOF
+  if [[ "${master}" == "false" ]]; then
```
Comment (mikedanese) [inline cluster/gce/util.sh]: Why is this needed?

File: cluster/gce/util.sh [hunk 12]
```diff
@@ -2444,6 +2698,45 @@ function create-nodes() {
...
+    this_mig_size=$((${instances_left} / (${NUM_WINDOWS_MIGS}-${i}+1)))
+    instances_left=$((instances_left-${this_mig_size}))
+
+    gcloud compute instance-groups managed \
+        create "${group_name}" \
+        --project "${PROJECT}" \
+        --zone "${ZONE}" \
+        --base-instance-name "${group_name}" \
+        --size "${this_mig_size}" \
+        --template "$template_name" || true;
```
Comment (mikedanese) [inline cluster/gce/util.sh]: Prefer braces:

https://google.github.io/styleguide/shell.xml?showone=Variable_expansion#Variable_expansion

File: cluster/gce/util.sh [hunk 13]
```diff
@@ -2444,6 +2698,45 @@ function create-nodes() {
...
+    gcloud compute instance-groups managed wait-until-stable \
+        "${group_name}" \
+        --zone "${ZONE}" \
+        --project "${PROJECT}" \
+        --timeout "${MIG_WAIT_UNTIL_STABLE_TIMEOUT}" || true;
+  done
+  wait
```
Comment (mikedanese) [inline cluster/gce/util.sh]: What are you waiting for? I don't see any subprocesses spawned.

**Review by mikedanese — COMMENTED** [review]
Ok, I looked at everything outside of  cluster/gce/win1803 and it's reasonable. I had a couple style nits. Can you cleanup git history?

File: cluster/gce/util.sh [hunk 14]
```diff
@@ -71,9 +80,29 @@ function set-node-image() {
...
 }
 
-set-node-image
+# Requires:
+#   WINDOWS_NODE_OS_DISTRIBUTION
+# Sets:
+#   WINDOWS_NODE_IMAGE_FAMILY
+#   WINDOWS_NODE_IMAGE_PROJECT
+function set-windows-node-image() {
+    WINDOWS_NODE_IMAGE_PROJECT="windows-cloud"
```
[the remark this answers is not in the document]
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 9 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done, thanks for the style guide link.

File: cluster/gce/util.sh (hunk 10 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 11 above)
Comment (pjh) [reply cluster/gce/util.sh]: I suspect there used to be something in this block but then it was removed. I've deleted it, cc @mtaufen in case there's some reason for it.

File: cluster/gce/util.sh (hunk 12 above)
Comment (pjh) [reply cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 13 above)
Comment (pjh) [reply cluster/gce/util.sh]: You're right. This was copied from create-nodes() just above... which also doesn't seem to create any subprocesses. (shrug)

Removed.

**Review by pjh — COMMENTED** [review]
Thanks for the comments @mikedanese!

**Comment by pjh** [comment]
> Ok, I looked at everything outside of cluster/gce/win1803 and it's reasonable. I had a couple style nits. Can you cleanup git history?

I've synced with head and squashed the commits back into one.

**Comment by michmike** [comment]
overall these changes lgtm. i won't approve, but i will provide my lgtm
/lgtm

**Comment by yujuhong** [comment]
/assign @mikedanese 
Thanks for the review!
